index.ts

```diff
@@ -1,59 +1,55 @@
 
-interface JestConfig {
-  transform: [match: string, transformerPath: string, options: Options][];
-  extensionsToTreatAsEsm?: string[]
-}
-
-interface JestTransformerOption {
-  config: JestConfig;
-  transformerConfig: Options;
-}
-
-/**
- * Loads closest package.json in the directory hierarchy
- */
-function loadClosestPackageJson(attempts = 1): Record<string, unknown> {
-  if (attempts > 5) {
-    throw new Error('Can\'t resolve main package.json file')
-  }
-  const mainPath = attempts === 1 ? './' : Array(attempts).join('../')
-  try {
-    return require(path.join(process.cwd(), mainPath, 'package.json'))
-  } catch (e) {
-    return loadClosestPackageJson(attempts + 1)
-  }
-}
-
-const packageConfig = loadClosestPackageJson()
-const isEsmProject = packageConfig.type === 'module'
-
-// Jest use the `vm` [Module API](https://nodejs.org/api/vm.html#vm_class_vm_module) for ESM.
-// see https://github.com/facebook/jest/issues/9430
-const supportsEsm = 'Module' in vm
-
-function createTransformer(swcTransformOpts?: Options) {
+function createTransformer(swcTransformOpts?: Options): Transformer {
   const computedSwcOptions = buildSwcTransformOpts(swcTransformOpts)
 
+  const cacheKeyFunction = getCacheKeyFunction([], [JSON.stringify(computedSwcOptions)])
+
   return {
-    process(src: string, filename: string, jestOptions: any) {
-      if (supportsEsm) {
-        set(computedSwcOptions, 'module.type', isEsm(filename, jestOptions) ? 'es6' : 'commonjs')
-      }
+    process(src, filename, jestOptions) {
+      set(computedSwcOptions, 'module.type', jestOptions.supportsStaticESM ? 'es6' : 'commonjs')
+
+      return transformSync(src, {
+        ...computedSwcOptions,
+        module: {
+          ...computedSwcOptions.module,
+          type: jestOptions.supportsStaticESM ? 'es6' : 'commonjs'
+        },
+        filename
+      })
+    },
+    processAsync(src, filename) {
+      return transform(src, {
+        ...computedSwcOptions,
+        module: {
+          ...computedSwcOptions.module,
+          // async transform is always ESM
+          type: 'es6'
+        },
+        filename
+      })
+    },
 
-      if (!computedSwcOptions.sourceMaps) {
-        set(computedSwcOptions, 'sourceMaps', 'inline')
-      }
+    getCacheKey(src, filename, ...rest){
+      // @ts-expect-error - type overload is confused
+      const baseCacheKey = cacheKeyFunction(src, filename, ...rest)
 
-      return transformSync(src, { ...computedSwcOptions, filename })
-    },
+      // @ts-expect-error - signature mismatch between Jest <27 og >=27
+      const options: TransformOptions = typeof rest[0] === 'string' ? rest[1] : rest[0]
 
-    getCacheKey: getCacheKeyFunction([], [JSON.stringify(computedSwcOptions)])
+      return crypto
+        .createHash('md5')
+        .update(baseCacheKey)
+        .update('\0', 'utf8')
+        .update(JSON.stringify({ supportsStaticESM: options.supportsStaticESM }))
```

Comment:
since the options are mutated based on this flag, we should have a separate cache key based on it in case the same file is transformed from both ESM and CJS